Support auto generation of device check

wenleix opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#56872 Support auto generation of device check**
* #56871 Disable device check for foreach kernels
* #56870 Add device_check place holder for functions
* #56869 Fix tensor device in test_kthvalue_overlap

Differential Revision: [D27986429](https://our.internmc.facebook.com/intern/diff/D27986429/)

Commit:
Support auto generation of device check



Differential Revision: [D27986429](https://our.internmc.facebook.com/intern/diff/D27986429/)

[ghstack-poisoned]

Commit:
rebase && fix minor issues on "Support auto generation of device check"


Differential Revision: [D27986429](https://our.internmc.facebook.com/intern/diff/D27986429/)

[ghstack-poisoned]

Commit:
fiix copysign&sparse on "Support auto generation of device check"


Differential Revision: [D27986429](https://our.internmc.facebook.com/intern/diff/D27986429/)

[ghstack-poisoned]

Commit:
fix cudnn&&nonzero on "Support auto generation of device check"


Differential Revision: [D27986429](https://our.internmc.facebook.com/intern/diff/D27986429/)

[ghstack-poisoned]

wenleix commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -143,6 +145,7 @@
     CUDA: _cudnn_rnn_flatten_weight
 
 - func: _cudnn_rnn(Tensor input, Tensor[] weight, int weight_stride0, Tensor? weight_buf, Tensor hx, Tensor? cx, int mode, int hidden_size, int proj_size, int num_layers, bool batch_first, float dropout, bool train, bool bidirectional, int[] batch_sizes, Tensor? dropout_state) -> (Tensor, Tensor, Tensor, Tensor, Tensor)
+  device_check: NoCheck  # TODO: explain this
```
`at::native::rnn_tanh` may or may not redispatch to `_cudnn_rnn` based on algorithm and build. And it might hit dispatch or kernel device check. Disable dispatch device check for consistent behavior.

wenleix commented on aten/src/ATen/core/op_registration/adaption.h:
```diff
@@ -59,5 +62,41 @@ check_tensor_options_and_extract_memory_format(
+TORCH_API void common_device_check_failure(optional<Device>& common_device, const at::Tensor& tensor, at::CheckedFrom methodName, at::CheckedFrom argName);
+
+inline void check_and_update_common_device(optional<Device>& common_device, const at::Tensor& tensor, at::CheckedFrom methodName, at::CheckedFrom argName) {
+  // TODO: Remove this check once all tensors are guaranteed to be defined
```
demonstrated by https://github.com/pytorch/pytorch/pull/56564. Should I open an issue to track this? @ezyang

ezyang replied:
Yes plz.

ezyang commented on test/test_linalg.py:
```diff
@@ -5127,13 +5127,13 @@ def test_householder_product_errors_and_warnings(self, device):
             wrong_device = 'cpu' if self.device_type != 'cpu' else 'cuda'
             out = torch.empty_like(reflectors).to(wrong_device)
-            with self.assertRaisesRegex(RuntimeError, "Expected result and input tensors to be on the same device"):
+            with self.assertRaisesRegex(RuntimeError, "Expected all tensors to be on the same device"):
```
Since you have manually fixed all of these, no further action is necessary. However, in the future, we should consider making error messages all use expect testing (see `test/test_dispatch.py` for example). These would then be updatable automatically using `EXPECTTEST_ACCEPT` without having to manually update the error strings in this case.

Commit:
rebase & address comment on "Support auto generation of device check"


Differential Revision: [D27986429](https://our.internmc.facebook.com/intern/diff/D27986429/)

[ghstack-poisoned]

Commit:
add README.md on "Support auto generation of device check"


Differential Revision: [D27986429](https://our.internmc.facebook.com/intern/diff/D27986429/)

[ghstack-poisoned]

Commit:
rebase && fix special xlogy on "Support auto generation of device check"


Differential Revision: [D27986429](https://our.internmc.facebook.com/intern/diff/D27986429/)

[ghstack-poisoned]
